### DISCUSSION TITLE
Enumerate adapters?

### DISCUSSION THREAD

**Discussion by toloudis** [gfx-rs/wgpu#2022] [Q&A] [answered]
Is there ever any intention of allowing more specific adapter selection by users?  For multi GPU systems and offscreen rendering it makes sense to have a very specific choice of adapter.  I saw an enumerate_adapters function but it looks very internal and not something usable from client code to actually select adapter from the results of enumerate_adapters.

**Comment by cwfitzgerald** [comment, marked as answer]
enumerate_adapters is exactly the function you should be using. You get a iterator of adapters which you can then use to get information `Adapter::features` `Adapter::limits` `Adapter::get_info` etc. From there you can decide on an adapter and call `request_device` on it.

We use it as part of our user space utilities: https://github.com/gfx-rs/wgpu/blob/master/wgpu/src/util/init.rs#L40-L59 selects an adapter based on its name.

**Comment by toloudis** [comment]
Perfect, thank you so much for that confirmation.

**Comment by almarklein** [comment]
That function is part of wgpu-rs. Are there plans to expose something similar in WebGPU and/or wgpu-native?

**Reply by cwfitzgerald** [reply]
wgpu-native it totally could be, WebGPU only ever has one adapter, so it wouldn't be available.

**Reply by kvark** [reply]
https://github.com/gpuweb/gpuweb/issues/2160#issuecomment-934705207 super fresh

**Reply by toloudis** [reply]
I completely understand the reasoning for WebGPU.   I was mainly thinking about working with native implementations.  I am interested in using the wgpu api for a cross platform graphics API and in particular the wgpu-py library.   And for the scientific volume rendering application I am developing my users have an interest in being able to select a GPU based on memory available, and they may have access to systems with multiple GPUs.

**Reply by cwfitzgerald** [reply]
Yeah, on native (or even on node/deno) there's no reason why this couldn't be exposed.
